File: packages/exchange-adapter/src/core/tools/Cache.ts
export class Cache<T> {
  private data: Map<string, { value: T; expiryTime: number }> = new Map();

  constructor(private defaultExpiryTime: number = 1000 * 60) { }

  set(
    key: string,
    value: T,
    expiryTime: number = this.defaultExpiryTime
  ): void {
    const expiryTimestamp = Date.now() + expiryTime;
    this.data.set(key, { value, expiryTime: expiryTimestamp });
  }

  get(key: string): T | undefined {
    // 在获取缓存值之前清理过期项
    this.cleanup();
    const item = this.data.get(key);

    if (item) {
      return item.value;
    } else {
      return undefined;
    }
  }
  clear(): void {
    this.data.clear();
  }
  private cleanup(): void {
    const now = Date.now();
    // 删除所有过期的缓存项
    this.data.forEach((item, key) => {
      if (item.expiryTime <= now) {
        this.data.delete(key);
      }
    });
  }
}

File: pkg/exchange-adapter/core/cache.go
package core

import (
	"sync"
	"time"
)

type cacheItem[T any] struct {
	value      T
	expiryTime time.Time
}

type Cache[T any] struct {
	mu                sync.Mutex
	data              map[string]cacheItem[T]
	defaultExpiryTime time.Duration
}

func NewCache[T any](defaultExpiry time.Duration) *Cache[T] {
	if defaultExpiry <= 0 {
		defaultExpiry = time.Minute
	}
	return &Cache[T]{
		data:              make(map[string]cacheItem[T]),
		defaultExpiryTime: defaultExpiry,
	}
}

func (c *Cache[T]) Set(key string, value T, expiry time.Duration) {
	if expiry <= 0 {
		expiry = c.defaultExpiryTime
	}
	c.mu.Lock()
	defer c.mu.Unlock()
	c.data[key] = cacheItem[T]{value: value, expiryTime: time.Now().Add(expiry)}
}

func (c *Cache[T]) Get(key string) (T, bool) {
	var zero T

	c.mu.Lock()
	defer c.mu.Unlock()

	c.cleanupLocked()

	item, ok := c.data[key]
	if !ok {
		return zero, false
	}
	return item.value, true
}

func (c *Cache[T]) Clear() {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.data = make(map[string]cacheItem[T])
}

func (c *Cache[T]) cleanupLocked() {
	now := time.Now()
	for k, item := range c.data {
		if !item.expiryTime.After(now) {
			delete(c.data, k)
		}
	}
}
